src/cargo/util/auth/mod.rs

```diff
@@ -552,20 +553,37 @@ fn credential_action(
         })?;
         match provider.perform(&registry, &action, &args[1..]) {
             Ok(response) => return Ok(response),
-            Err(cargo_credential::Error::UrlNotSupported) => {}
-            Err(cargo_credential::Error::NotFound) => any_not_found = true,
-            e => {
-                return e.with_context(|| {
-                    format!(
-                        "credential provider `{}` failed action `{action}`",
-                        args.join(" ")
-                    )
-                })
-            }
+            Err(e) => match e.kind() {
+                cargo_credential::ErrorKind::UrlNotSupported => {
```

Comment:
I'm not sure that the `UrlNotSupported` error makes sense for "non-fatal" errors. It was intended to indicate that the provider only supports some registry urls (like a provider only for a specific type of registry). It then grew to also be used for the platform-specific providers on the wrong platform - and that still made some sense since they are providers that don't work with any registry.

This use case of non-fatal (continuable) error is much further from that intent. However, I'm not sure what the right solution is.

Here are some other options:

1. Make all errors "non-fatal" and continue trying providers for all error types
2. Keep the existing behavior, but add an option in the global credential providers config to mark a provider as "optional"
3. Add a new error type of `OtherContinuable` that that works as you've implemented it here.

Reply:
Talked about this in office hours

Add an alternative to `Other` that is `Backtrack`
- `Backtrack` will need an MSRV note in docs
- Maybe add an alias for `Other` that is `Critical`
  - Be careful about forcing MSRV bump

As for whether every error could have a message, this was a bit unclear.
- `url-not-supported`: doesn't need it
- `not-found`: doesn't need it
- `operation-not-supported`: could benefit from it because a login failure on a provider that doesn't support it could give a better message saying why.

In the end, we decided that since the code was written, we should just move forward with a message on each one.

Reply:
For message printing:
- On error (everything backtracked):
  - Print every provider that backtracked and the message
- On success (a provider matched)
  - In extra verbose, print the providers that backtracked and the message

Reply:
Since we're doing a breaking change, we should also do #13615

Reply:
@arlosi will do that breaking change after this is merged but within the same release